active-rfcs/0000-attr-fallthrough.md

```diff
@@ -0,0 +1,220 @@
+
+- `inheritAttrs: false` now affects `class` and `style`.
+
+- `this.$attrs` now contains everything passed to the component minus those explicitly declared as props, including `class`, `style`, and `v-on` listeners. `this.$listeners` is removed.
+
+- Functional components attribute fallthrough behavior adjusted:
+  - With explicit `props` declaration: full fallthrough like stateful components.
+  - With no `props` declaration: only fallthrough for `class`, `style` and `v-on` listeners.
+
+# Motivation
+
+In Vue 2.x, components have an implicit attributes fallthrough behavior. Any attribute passed to a component that is not declared as a prop by the component, is considered an **extraneous attribute**. Fore example:
+
+``` html
+<MyComp id="foo"/>
+```
+
+If `MyComp` didn't declare a prop named `id`, then the `id` is considered an extraneous attribute and will implicitly be applied to the root node of `MyComp`.
+
+This behavior is very convenient when tweaking layout styling between parent and child (by passing on `class` and `style`), or applying a11y attributes to child components.
+
+This behavior can be disabled with `inheritAttrs: false`, where the user expects to explicitly control where the attributes should be applied. These extraneous attributes are exposed in an instance property: `this.$attrs`.
+
+There are a number of inconsistencies and issues in the 2.x behavior:
+
+- `inheritAttrs: false` does not affect `class` and `style`.
+
+- Implicit fallthrough does not apply for event listeners, leading to the need for `.native` modifier if the user wish to add a native event listener to the child component root.
+
+- `class`, `style` and `v-on` listeners are not included in `$attrs`, making it cumbersome for a higher-order component (HOC) to properly pass everything down to a nested child component.
+
+- Functional components have no implicit attrs fallthrough behavior.
+
+In 3.x, we are also introducing Fragments (multiple root nodes in a component template), which require additional considerations on the behavior.
+
+# Detailed design
+
+## `v-on` Listener Fallthrough
+
+With the following usage:
+
+```html
+<MyButton @click="hello" />
+```
+
+- In v2, the `@click` will only register a component custom event listener. To attach a native listener to the root of `MyButton`, `@click.native` is needed.
+
+- In v3, the `@click` listener will fallthrough and register a native click listener on the root of `MyButton`. This means component authors no longer need to proxy native events to custom events in order to support `v-on` usage without the `.native` modifier. In fact, the `.native` modifier will be removed altogether.
+
+  Note this may result in unnecessary registration of native event listeners when the user is only listening to component custom events, which we discuss below in [Unresolved Questions](#unresolved-questions).
```

Comment:
If `MyButton` is authored in that way would it produce 2 click events with implicit event listeners fallthrough?

```html
<button @click="$emit('click', $event)">Click me</button>
```

Reply:
Yep, the whole point is you don't proxy events anymore.